### PR TITLE
Strategy Pattern

### PR DESCRIPTION
**Pull request by tonilopezmr**
I have added Strategy pattern with IVA, General, Reduced and Super Reduced. 👍 

And @Cotel very crack `scanner.nextInt()` this is a big fail :1st_place_medal:

### CODE REVIEW

**Comment by Cotel** [comment]
Vais a flipar pepinillos...

*ehem* I mean, you will love the last commit 👍

File: src/main/kotlin/Strategy/strategies/GeneralStrategy.kt [hunk 1]
```diff
@@ -1,11 +1,3 @@
...
-import Strategy.base.IvaStrategy
-
-class GeneralStrategy : IvaStrategy {
-
-  override fun applyIva(cost: Double): Double {
-    return cost + cost * 0.21
-  }
-
-}
+val GeneralStrategy : (Double) -> Double = { it + it * 0.21 }
```
Comment (Cotel) [inline src/main/kotlin/Strategy/strategies/GeneralStrategy.kt]: We don't really need to make a file just for this, but it's clearer for comparing with traditional way of writing the pattern.

File: src/main/kotlin/Strategy/strategies/ReducedStrategy.kt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package Strategy.strategies
+
+val ReducedStrategy : (Double) -> Double = { it + it * 0.10 }
```
Comment (frnlpz) [inline src/main/kotlin/Strategy/strategies/ReducedStrategy.kt]: What's `it` referring to here?

Comment (Cotel) [inline src/main/kotlin/Strategy/strategies/ReducedStrategy.kt]: `it` is a lambda here.

In kotlin you can make labda funtions like this: 

```kotlin
val numbers = listOf(1,2,3,4)
numbers.map { number -> number * number } // returns [1, 4, 9, 16]
```

We can omit `number` parameter definition and call it `it` instead.

```kotlin
numbers.map { it * it }
```

In our example, `ReducedStrategy` is a **high order function** which receives a _Double_ and returns a _Double_. `it` is referencing that _Double_ parameter.

Comment (frnlpz) [inline src/main/kotlin/Strategy/strategies/ReducedStrategy.kt]: What an amazing property of Kotlin. I'm so flabbergasted

File: src/main/kotlin/Command/base/Cart.kt [hunk 3]
```diff
@@ -95,17 +94,17 @@ open class Cart(private val scanner: Scanner,
...
+      3 -> ivaStrategy = SuperReducedStrategy
+      else -> ivaStrategy = GeneralStrategy
     }
   }
 
   fun showPrice() {
     chooseStrategy()
 
-    println("> Your order is: ${ivaStrategy.applyIva(noodles.calculateCost())} $")
+    println("> Your order is: ${ivaStrategy.invoke(noodles.calculateCost())} $")
```
Comment (tonilopezmr) [inline src/main/kotlin/Command/base/Cart.kt]: This is not necessary, I don't like it you can remove invoke 💯

File: src/main/kotlin/Strategy/strategies/ReducedStrategy.kt (hunk 2 above)
Comment (tonilopezmr) [inline src/main/kotlin/Strategy/strategies/ReducedStrategy.kt]: Java8 has it too @frnlpz 👍

File: src/main/kotlin/Strategy/strategies/SuperReducedStrategy.kt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+package Strategy.strategies
+
+// Alternative way
+val SuperReducedStrategy = { cost: Double -> cost + cost * 0.04 }
```
Comment (tonilopezmr) [inline src/main/kotlin/Strategy/strategies/SuperReducedStrategy.kt]: l3l :1st_place_medal:

File: src/main/kotlin/Command/base/Cart.kt [hunk 5]
```diff
@@ -104,7 +104,7 @@ open class Cart(private val scanner: Scanner,
   fun showPrice() {
     chooseStrategy()
 
-    println("> Your order is: ${ivaStrategy.invoke(noodles.calculateCost())} $")
+    println("> Your order is: ${ivaStrategy(noodles.calculateCost())} $")
```
Comment (Cotel) [inline src/main/kotlin/Command/base/Cart.kt]: :+1:

File: src/test/kotlin/Strategy/strategies/IVAStrategyTest.kt [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import Strategy.strategies.*
+import org.hamcrest.CoreMatchers
+import org.junit.Assert.*
+import org.junit.Test
+
+class IVAStrategyTest{
+
+  @Test
+  fun `check that general iva is applied correctly`(){
+    assertThat(GeneralStrategy(2.0), CoreMatchers.`is`(2+2*0.21))
```
Comment (tonilopezmr) [inline src/test/kotlin/Strategy/strategies/IVAStrategyTest.kt]: Perfect 👍

Comment (Cotel) [inline src/test/kotlin/Strategy/strategies/IVAStrategyTest.kt]: Shouln't the result be hardcoded? I mean, the strategy is making that same operation. It won't fail never. We should write the exact result for (2 + 2 * 0.21) for seeing if the calculus is returning what we expect. Am I right?

Comment (tonilopezmr) [inline src/test/kotlin/Strategy/strategies/IVAStrategyTest.kt]: You mean replace `2+2*0,21` for `2,42`  ??

Comment (Cotel) [inline src/test/kotlin/Strategy/strategies/IVAStrategyTest.kt]: Exactly

**Comment by Cotel** [comment]
This close #6 btw

File: src/test/kotlin/Strategy/strategies/IVAStrategyTest.kt (hunk 6 above)
Comment (Cotel) [inline src/test/kotlin/Strategy/strategies/IVAStrategyTest.kt]: Can we change it or we merge already?

Comment (tonilopezmr) [inline src/test/kotlin/Strategy/strategies/IVAStrategyTest.kt]: @frnlpz HEEEYYY???

**Comment by tonilopezmr** [comment]
👍

**Pull request merged**